Support for Curve 25519 - Algorithm Ed25519

angelaizg opened the pull request:
#### 49d4ef6da92f266244e072be6be48f29c4043dca
<pre>
Support for Curve 25519 - Algorithm Ed25519
https://bugs.webkit.org/show_bug.cgi?id=246145
rdar://100401588

Reviewed by NOBODY (OOPS!).

Adding support for the Curve 25519 (doing the necessary modifications to be able to work with such type  of curve), specifically the implementation of the Ed25519 algorithm, as in https://www.rfc-editor.org/rfc/rfc7748. The current patch adds support  to generate keys, import, export (in raw format for now), sign and verify and tests that make use of such operations. Following the specs in https://wicg.github.io/webcrypto-secure-curves/#ed25519.

* LayoutTests/crypto/subtle/ed25519-generate-export-key-raw-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-generate-export-key-raw.html: Added.
* LayoutTests/crypto/subtle/ed25519-generate-key-sing-verify-Curve25519-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-generate-key-sing-verify-Curve25519.html: Added.
* LayoutTests/crypto/subtle/ed25519-import-raw-key-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-import-raw-key.html: Added.
* LayoutTests/crypto/subtle/ed25519-serialize-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-serialize.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h: Added.
(cccurve25519_make_priv):
(cccurve25519_make_pub):
(cccurve25519_make_key_pair):
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::read):
* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
* Source/WebCore/crypto/CryptoAlgorithmParameters.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::normalizeCryptoAlgorithmParameters):
(WebCore::isSupportedExportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmECDH.cpp:
(WebCore::CryptoAlgorithmECDH::generateKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmECDSA.cpp:
(WebCore::CryptoAlgorithmECDSA::generateKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp: Copied from Source/WebCore/crypto/algorithms/CryptoAlgorithmECDSA.cpp.
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
(WebCore::CryptoAlgorithmEd25519::create):
(WebCore::CryptoAlgorithmEd25519::identifier const):
(WebCore::CryptoAlgorithmEd25519::generateKey):
(WebCore::CryptoAlgorithmEd25519::sign):
(WebCore::CryptoAlgorithmEd25519::verify):
(WebCore::CryptoAlgorithmEd25519::importKey):
(WebCore::CryptoAlgorithmEd25519::exportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h: Added.
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp: Copied from Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp.
(WebCore::gcryptSign):
(WebCore::gcryptVerify):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmRegistryGCrypt.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/gcrypt/CryptoKeyECGCrypt.cpp:
(WebCore::curveName):
(WebCore::curveIdentifier):
(WebCore::curveSize):
(WebCore::curveUncompressedFieldElementSize):
* Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
(WebCore::toNamedCurve):
(WebCore::CryptoKeyEC::exportJwk const):
(WebCore::CryptoKeyEC::namedCurveString const):
(WebCore::CryptoKeyEC::isValidECAlgorithm):
(WebCore::CryptoKeyEC::algorithm const):
* Source/WebCore/crypto/keys/CryptoKeyEC.h:
* Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp:
(WebCore::CryptoAlgorithmECDH::platformDeriveBits):
* Source/WebCore/crypto/mac/CryptoAlgorithmECDSAMac.cpp:
(WebCore::signECDSA):
(WebCore::verifyECDSA):
* Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp: Added.
(WebCore::signEd25519):
(WebCore::verifyEd25519):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/mac/CryptoAlgorithmHKDFMac.cpp:
* Source/WebCore/crypto/mac/CryptoAlgorithmRegistryMac.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
(WebCore::keySizeInBitsFromNamedCurve):
(WebCore::CryptoKeyEC::keySizeInBits const):
(WebCore::CryptoKeyEC::platformSupportedCurve):
(WebCore::CryptoKeyEC::platformGeneratePair):
(WebCore::CryptoKeyEC::platformImportRaw):
(WebCore::CryptoKeyEC::platformExportRaw const):
(WebCore::CryptoKeyEC::platformImportJWKPublic):
(WebCore::CryptoKeyEC::platformImportJWKPrivate):
(WebCore::CryptoKeyEC::platformAddFieldElements const):
(WebCore::getOID):
(WebCore::CryptoKeyEC::platformImportSpki):
(WebCore::CryptoKeyEC::platformExportSpki const):
(WebCore::CryptoKeyEC::platformImportPkcs8):
(WebCore::CryptoKeyEC::platformExportPkcs8 const):
* Source/WebCore/crypto/openssl/CryptoAlgorithmEd25519OpenSSL.cpp: Copied from Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp.
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/openssl/CryptoAlgorithmRegistryOpenSSL.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h: Copied from Source/WebCore/crypto/CryptoAlgorithmIdentifier.h.
* Source/WebCore/crypto/parameters/Ed25519Params.idl: Copied from Source/WebCore/crypto/CryptoAlgorithmIdentifier.h.
* Source/WebCore/css/StylePropertyShorthand.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cfb9135493dd76c002bbb5f608c5d077fe25198

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9474 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33378 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109632 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169849 "Failed to checkout and rebase branch from PR 5026") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104304 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/15 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92717 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107512 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106089 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7982 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91126 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34519 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/89895 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77474 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3220 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24035 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3209 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/147 "Found 1 new test failure: http/tests/security/contentSecurityPolicy/source-list-parsing-05.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9331 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43525 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5029 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/crypto/parameters/Ed25519Params.idl:
```diff
@@ -0,0 +1,31 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+ typedef (object or DOMString) HashAlgorithmIdentifier;
+[
```
If you add an idl file, you need to add it to Source/WebCore/CMakeLists.txt so that it compiles in GTK/WPE

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h:
```diff
@@ -0,0 +1,43 @@
+#include <JavaScriptCore/Strong.h>
+#include <variant>
+
+
```
one line to remove

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -43,6 +45,7 @@ static constexpr unsigned char Secp256r1[] = {0x06, 0x08, 0x2a, 0x86, 0x48, 0xce
 static constexpr unsigned char Secp384r1[] = {0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x22};
 // OID secp521r1 1.3.132.0.35
 static constexpr unsigned char Secp521r1[] = {0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x23};
+static constexpr unsigned char curveEd25519[] = { 0x06, 0x09, 0x2b, 0x06, 0x01, 0x04, 0x01, 0xda, 0x47, 0x0f, 0x01 };
```
The other vars seem to have the first letter highcased

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,51 +299,53 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
     if (keyData.size() < index + 1)
         return nullptr;
-    index += bytesUsedToEncodedLength(keyData[index]) + 1; // Read length
+    index += bytesUsedToEncodedLength(keyData[index])+1; // Read length
```
These changes are probably unrelated.
Might another commit that was squashed. You probably want to remove them.

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,45 @@
+    CryptoAlgorithmEd25519() = default;
+    CryptoAlgorithmIdentifier identifier() const final;
+    
+    
```
Extra line.

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,76 @@
+    (void) keyLengthInBytes;
+    (void) hash;
+    (void) key;
+    Vector<uint8_t> pk = std::get<Vector<uint8_t>>(key);
```
If the key was typed, we would directly get a Vector and not a variant here.

youennf replied:
Extra copy not needed.

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,76 @@
+//  WebCore
+//
+//  Created by Ángela Izquierdo García on 14/9/22.
+//
```
Please use the regular WebKit copyright.

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,76 @@
+//
+
+#include <stdio.h>
+ /* CryptoAlgorithmEd25519 */
```
Please remove these two lines.

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,76 @@
+
+#include <string.h>
+
+
```
Extra lines

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,76 @@
+
+#include "CryptoDigestAlgorithm.h"
+#include <wtf/text/Base64.h>
+#include <wtf/Assertions.h>
```
Missing extra line.

angelaizg replied:
where should the extra line be?

aperezdc replied:
I think in this case there is no line missing, but there are indeed some issues with the how the includes section is currently formatted. You can see the results from the style checker bot here: https://ews-build.webkit.org/#/builders/6/builds/91635

Tip: If you want to run the style checker locally before updating the PR, you can run `Tools/Scripts/check-webkit-style -g HEAD` (assuming that you want to check the most recent commit in your work branch, of course you can pass a diffeernt commit-id to `-g` if needed).

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,76 @@
+{
+    (void) keyLengthInBytes;
+    (void) hash;
+    (void) key;
```
Why these lines?

angelaizg replied:
As they are parameters not used passed to the function it was giving a compiling error but the parameters passed should remain the same so that's why I wrote it that way

aperezdc replied:
We have an `UNUSED()` macro for this, or better: C++ allows omitting the parameter names from the function definition (leaving only the types).

aperezdc commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp.rej:
```diff
@@ -0,0 +1,19 @@
+diff a/Source/WebCore/bindings/js/SerializedScriptValue.cpp b/Source/WebCore/bindings/js/SerializedScriptValue.cpp	(rejected hunks)
```
This file should not be in the patch, please remove it before landing.

aperezdc commented on Source/WebCore/crypto/CryptoAlgorithmIdentifier.h.rej:
```diff
@@ -0,0 +1,9 @@
+diff a/Source/WebCore/crypto/CryptoAlgorithmIdentifier.h b/Source/WebCore/crypto/CryptoAlgorithmIdentifier.h	(rejected hunks)
```
Same for this one. Please make sure there are no `.rej`, `.orig`, etc. files.

youennf commented on LayoutTests/crypto/subtle/ed25519-generate-export-key-raw.html:
```diff
@@ -0,0 +1,40 @@
+
+var keyPair;
+debug("Generating a key pair...");
+crypto.subtle.generateKey(algorithmKeyGen, extractable, ["sign", "verify"]).then(function(result) {
```
It would be nice if these tests were written as WPT tests (using testharness.js), since they can be used by other browsers.

angelaizg replied:
I have undone the changes on the WPT and I think that would be better as a follow-up patch

youennf commented on LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/generateKey/failures.js:
```diff
@@ -94,7 +94,7 @@ function run_test(algorithmNames) {
             });
         } else if (algorithmName.toUpperCase().substring(0, 2) === "EC") {
-            ["P-512", "Curve25519"].forEach(function(curveName) {
+            ["P-512"].forEach(function(curveName) {
```
Can you explain the reason for this change?

angelaizg replied:
Now that the curve is implemented there is no need to address it as a not implemented curve and do such checks

youennf replied:
Then we should not change this file?

youennf commented on Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
```diff
@@ -74,14 +74,14 @@
 		416E995323DAE6BE00E871CB /* AudioToolboxSoftLink.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 416E995123DAE6BD00E871CB /* AudioToolboxSoftLink.cpp */; };
 		41E1F344248A6A000022D5DE /* VideoToolboxSoftLink.cpp in Sources */ = {isa = PBXBuildFile; fileRef = 416E995523DAEFF700E871CB /* VideoToolboxSoftLink.cpp */; };
 		41E9EE2328BB57A0006A2298 /* MediaExperienceAVSystemControllerSPI.h in Headers */ = {isa = PBXBuildFile; fileRef = 41E9EE2228BB57A0006A2298 /* MediaExperienceAVSystemControllerSPI.h */; };
-		41E9EE2628BCA19E006A2298 /* SBSStatusBarSPI.h in Headers */ = {isa = PBXBuildFile; fileRef = 41E9EE2528BCA19E006A2298 /* SBSStatusBarSPI.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
I do not think SBSStatusBarSPI.h should be changed

youennf commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,103 @@
+
+#pragma once
+
+extern "C" {
```
We usually use WTF_EXTERN_C_BEGIN.

youennf commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -2814,6 +2818,9 @@ class CloneDeserializer : CloneBase {
             break;
+        case CryptoAlgorithmIdentifierTag::Ed25519:
+            result  = CryptoAlgorithmIdentifier::Ed25519;
+            break;
```
It might be good to add a test that validates Ed25519 key serialisation works.

angelaizg replied:
What do you mean about this?

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -96,7 +98,6 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
     if (std::holds_alternative<String>(algorithmIdentifier)) {
         auto newParams = Strong<JSObject>(vm, constructEmptyObject(&state));
         newParams->putDirect(vm, Identifier::fromString(vm, "name"_s), jsString(vm, std::get<String>(algorithmIdentifier)));
-        
```
not needed

angelaizg replied:
what is exactly what is not needed ?

youennf replied:
We usually do not remove lines except if changing the file.
There is no need to change SubtleCrypto.cpp in this patch.

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -106,8 +107,10 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
+    
     if (UNLIKELY(!identifier))
         return Exception { NotSupportedError };
+    
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -106,8 +107,10 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
     RETURN_IF_EXCEPTION(scope, Exception { ExistingExceptionError });
 
     auto identifier = CryptoAlgorithmRegistry::singleton().identifier(params.name);
+    
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -163,6 +166,13 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
+            RETURN_IF_EXCEPTION(scope, Exception { ExistingExceptionError });
+            result = makeUnique<CryptoAlgorithmEd25519Params>(params);
+            break;
+
```
line not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -194,6 +204,7 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
             result = makeUnique<CryptoAlgorithmRsaKeyGenParams>(params);
             break;
         }
+
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -252,6 +264,7 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
             result = makeUnique<CryptoAlgorithmEcdhKeyDeriveParams>(params);
             break;
         }
+
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -366,7 +380,6 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
         }
         break;
     }
-
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -853,6 +867,7 @@ void SubtleCrypto::deriveKey(JSC::JSGlobalObject& state, AlgorithmIdentifier&& a
 
 void SubtleCrypto::deriveBits(JSC::JSGlobalObject& state, AlgorithmIdentifier&& algorithmIdentifier, CryptoKey& baseKey, unsigned length, Ref<DeferredPromise>&& promise)
 {
+    
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -936,7 +951,6 @@ void SubtleCrypto::exportKey(KeyFormat format, CryptoKey& key, Ref<DeferredPromi
         promise->reject(Exception { NotSupportedError });
         return;
     }
-
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -1073,9 +1087,7 @@ void SubtleCrypto::wrapKey(JSC::JSGlobalObject& state, KeyFormat format, CryptoK
 void SubtleCrypto::unwrapKey(JSC::JSGlobalObject& state, KeyFormat format, BufferSource&& wrappedKeyBufferSource, CryptoKey& unwrappingKey, AlgorithmIdentifier&& unwrapAlgorithmIdentifier, AlgorithmIdentifier&& unwrappedKeyAlgorithmIdentifier, bool extractable, Vector<CryptoKeyUsage>&& keyUsages, Ref<DeferredPromise>&& promise)
 {
     auto wrappedKey = copyToVector(WTFMove(wrappedKeyBufferSource));
-
```
not needed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -1073,9 +1087,7 @@ void SubtleCrypto::wrapKey(JSC::JSGlobalObject& state, KeyFormat format, CryptoK
     auto wrappedKey = copyToVector(WTFMove(wrappedKeyBufferSource));
-
     bool isDecryption = false;
-
```
not needed

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,64 @@
+    static constexpr CryptoAlgorithmIdentifier s_identifier = CryptoAlgorithmIdentifier::Ed25519;
+    static Ref<CryptoAlgorithm> create();
+    
+    
```
Line unneeded

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,64 @@
+    CryptoAlgorithmEd25519() = default;
+    CryptoAlgorithmIdentifier identifier() const final;
+    
+    void generateKey(const CryptoAlgorithmParameters& , bool extractable, CryptoKeyUsageBitmap usages, KeyOrKeyPairCallback&& , ExceptionCallback&& , ScriptExecutionContext&);
```
final?

angelaizg replied:
what do you mean ?

cdumez replied:
https://en.cppreference.com/w/cpp/language/final

angelaizg replied:
all the headers of algorithms are declared that way, I just followed the same pattern, shall I change it the?

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,64 @@
+    
+    static ExceptionOr<Vector<uint8_t>> platformSign(const CryptoAlgorithmEd25519Params&, const CryptoKeyEC&, const Vector<uint8_t>&);
+    static ExceptionOr<bool> platformVerify(const CryptoAlgorithmEd25519Params&, const CryptoKeyEC&, const Vector<uint8_t>&, const Vector<uint8_t>&);
+    
```
unneeded line

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
```diff
@@ -57,8 +63,11 @@ CryptoKeyEC::CryptoKeyEC(CryptoAlgorithmIdentifier identifier, NamedCurve curve,
 {
     // Only CryptoKeyEC objects for supported curves should be created.
     ASSERT(platformSupportedCurve(curve));
+
```
unneeded line

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
```diff
@@ -57,8 +63,11 @@ CryptoKeyEC::CryptoKeyEC(CryptoAlgorithmIdentifier identifier, NamedCurve curve,
 }
 
+
+
```
unneeded line

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
```diff
@@ -138,7 +147,6 @@ ExceptionOr<Vector<uint8_t>> CryptoKeyEC::exportRaw() const
 {
     if (type() != CryptoKey::Type::Public)
         return Exception { InvalidAccessError };
-
```
unneeded change

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
```diff
@@ -198,6 +209,9 @@ String CryptoKeyEC::namedCurveString() const
         return String(P521);
+    case NamedCurve::Curve25519:
+        return String(Curve25519);
+
```
unneeded line

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmECDSAMac.cpp:
```diff
@@ -37,7 +37,7 @@
 
 static ExceptionOr<Vector<uint8_t>> signECDSA(CryptoAlgorithmIdentifier hash, const PlatformECKey key, size_t keyLengthInBytes, const Vector<uint8_t>& data)
-{
+{   
```
unneeded change

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+namespace WebCore {
+static ExceptionOr<Vector<uint8_t>> signEd25519(CryptoAlgorithmIdentifier, const PlatformECKey key, size_t , const Vector<uint8_t>& data)
+{
+    Vector<uint8_t> sk = std::get<Vector<uint8_t>>(key);
```
This is causing an unneeded extra copy.
You can probably use auto&

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+    ccec25519signature newSignature = { };
+    cced25519_sign(di, newSignature, data.size(), data.data(), pk, sk.data());
+    Vector<uint8_t> sg(newSignature, 64);
+    return WTFMove( sg);
```
```suggestion
    return WTFMove(sg);
```

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+    ccec25519pubkey pk;
+    const struct ccdigest_info* di = ccsha512_di();
+    cced25519_make_pub(di, pk, sk.data());
+    Vector<uint8_t> publicKey(pk, 32);
```
Why do we need publicKey?

angelaizg replied:
It is required by the function of the library

youennf replied:
publicKey variable does not seem to be used anywhere.

cdumez replied:
As Youenn said, the publicKey variable is unused and should be dropped (or you should make use of it).

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+    cced25519_sign(di, newSignature, data.size(), data.data(), pk, sk.data());
+    Vector<uint8_t> sg(newSignature, 64);
+    return WTFMove( sg);
+
```
unneeded line

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+    int result = cced25519_verify(di, data.size(), data.data(), signature.data(), pk.data());
+    if (result)
+        return false;
+    return true;
```
You can use
return !cced25519_verify(di, data.size(), data.data(), signature.data(), pk.data())

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+}
+
+
+
```
Extra lines unneeded

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,90 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+            return options.size()*8;
+    });
+
+
```
extra lines

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,90 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+        return CCECGetKeySize(options.get());
+        },
+        [](const Vector<uint8_t>& options) -> int {
+            return options.size()*8;
```
```suggestion
            return options.size() * 8;
```

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,90 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+}
+PlatformECKey CryptoKeyEC::platformKey() const
+{
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) -> PlatformECKey {
```
This creates a lot of switchOn.
Should we not use a separate class for E25519 keys?

angelaizg replied:
The Ed25519 is a type of elliptic curve so I think it is better if we treat it as such because most of their procedures are the same conceptuall

youennf replied:
We use variants if we can share some code.
Can you tell which code is actually shared?
If we are not sharing any code between the two, it might make the code more complex to understand without real benefits.

cdumez replied:
Extra space after `WTF::switchOn(`

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,82 @@
+    const struct ccdigest_info* di = ccsha512_di();
+    cced25519_make_pub(di, pk, sk.data());
+    Vector<uint8_t> publicKey(pk, 32);
+    ccec25519signature newSignature = { };
```
We probably do not need = { }

mcatanzaro reviewed (changes requested):
You have 243 unrelated commits in this pull request, modifying 2299 files, which is why 19 different people have been added to this review as code owners. Needless to say, we can't review this. You'll need to rebase your branch using `git rebase -i origin/main` to fix this.

youennf commented on LayoutTests/crypto/subtle/ed25519-import-raw-key.html:
```diff
@@ -0,0 +1,41 @@
+jsTestIsAsync = true;
+
+//var rawKey = hexStringToUint8Array("04dc08d9bf603333eab1ad29cef41a6203ab6ecce03c5a4a9bf5771a3fb9f971d89a09664bfb87c61199b3453220eadec714c49ad1e24bf5d7ec5bddeca6420893");
+//var rawKey =  new Uint8Array([43, 126, 208, 188, 119, 149, 105, 74, 180, 172, 211, 89, 3, 254, 140, 215, 216, 15, 106, 28, 134, 136, 166, 195, 65, 68, 9, 69, 117, 20, 161, 69, 120, 85, 187, 178, 25, 227, 10, 27, 238, 168, 254, 134, 144, 130, 217, 159, 200, 40, 47, 144, 80, 208, 36, 229, 158, 175, 7, 48, 186, 157, 183, 10]);
```
Why this commented line?

angelaizg replied:
Just tried different formats, going to update it

youennf commented on LayoutTests/imported/w3c/resources/import-expectations.json:
```diff
@@ -51,6 +51,7 @@
     "web-platform-tests/IndexedDB": "import", 
     "web-platform-tests/IndexedDB/resources": "import", 
     "web-platform-tests/WebCryptoAPI": "import", 
+    "web-platform-tests/WebCryptoAPI/sign_verify": "import", 
```
Not needed, let's remove it.

angelaizg replied:
why is that?

youennf commented on LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/sign_verify/eddsa.https.any.worker-expected.txt:
```diff
@@ -0,0 +1,39 @@
+
+PASS setup
+FAIL importVectorKeys step: Ed25519 verification assert_unreached: importVectorKeys failed for//// Ed25519SHA-512Curve25519. Message: ''Data provided to an operation does not meet requirements'' Reached unreachable code
```
Why are we failing this test?

angelaizg replied:
Currently we are working with the raw format only, the other ones are not expected to work. Nevertheless, Ed25519-raw  are all passing

cdumez commented on LayoutTests/crypto/subtle/ed25519-generate-key-sing-verify-Curve25519.html:
```diff
@@ -0,0 +1,55 @@
+    var arrayBuffer = crypto.subtle.exportKey("raw",keyPair.privateKey);
+    var base64String = btoa(String.fromCharCode.apply(null, new Uint8Array(arrayBuffer)));
+    
+    /*const exported = await window.crypto.subtle.exportKey(
```
We shouldn't leave commented out code.

cdumez commented on LayoutTests/crypto/subtle/ed25519-generate-key-sing-verify-Curve25519.html:
```diff
@@ -0,0 +1,55 @@
+    shouldBe("signature.byteLength", "64");
+
+    debug("Verifying");
+    //return crypto.subtle.verify(ed25519Params, keyPair.publicKey, signature, data);
```
No commented out code.

cdumez commented on LayoutTests/imported/w3c/resources/import-expectations.json:
```diff
@@ -489,4 +489,4 @@
     "webrtc-platform-tests/webrtc": "import"
-}
\ No newline at end of file
+}
```
Please revert changes to this file.

cdumez commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,107 @@
+#define    CCSHA512_STATE_SIZE   64
+extern const struct ccdigest_info ccsha512_ltc_di;
+
+
```
Too many blank lines.

cdumez commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,107 @@
+    sk[31] &= 127;
+    sk[31] |= 64;
+}
+inline void cccurve25519_make_pub(ccec25519pubkey pk, const ccec25519secretkey sk)
```
Missing blank line before.

angelaizg replied:
where ?

cdumez commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,107 @@
+
+int cced25519_make_pub(const struct ccdigest_info *, ccec25519pubkey pk, const ccec25519secretkey sk);
+
+
```
extra blank line.

cdumez commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -92,23 +94,22 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
     VM& vm = state.vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
-
+    
```
Bad spacing change.

cdumez commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -92,23 +94,22 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
-
     auto& value = std::get<JSC::Strong<JSC::JSObject>>(algorithmIdentifier);
-
+    
```
Bad spacing change, please undo.

cdumez commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (C) 2017 Apple Inc. All rights reserved.
```
Why 2017?

cdumez commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,207 @@
+#include "config.h"
+#include "CryptoAlgorithmEd25519.h"
+
+
```
Too many blank lines.

cdumez commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2017 Apple Inc. All rights reserved.
```
Why 2017?

cdumez commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,58 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
```
Too many blank lines.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
```diff
@@ -130,15 +137,15 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::importPkcs8(CryptoAlgorithmIdentifier identifie
     if (!namedCurve || !platformSupportedCurve(*namedCurve))
         return nullptr;
-
+    
```
Bad spacing change, please undo all these.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -33,14 +33,16 @@
 #include "CommonCryptoUtilities.h"
+#include <wtf/Vector.h>
+
+typedef std::variant<CCECCryptorRef, Vector<uint8_t>>PlatformECKey;
```
missing space before `PlatformECKey`. Also, we prefer `using` the typedef nowadays:
`using PlatformECKey = std::variant<CCECCryptorRef, Vector<uint8_t>>;`

angelaizg replied:
done

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2017 Apple Inc. All rights reserved.
```
Why 2017?

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,75 @@
+#include <wtf/text/Base64.h>
+
+namespace WebCore {
+static ExceptionOr<Vector<uint8_t>> signEd25519(CryptoAlgorithmIdentifier, const PlatformECKey key, size_t , const Vector<uint8_t>& data)
```
Missing blank line before this.

cdumez replied:
Also, you have an extra space after `size_t`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoAlgorithmRegistryMac.cpp:
```diff
@@ -48,6 +49,7 @@
 #include "CryptoAlgorithmSHA384.h"
 #include "CryptoAlgorithmSHA512.h"
 
+
```
Too many blank lines here.

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
-    int result = CCECGetKeySize(m_platformKey.get());
-    return result ? result : 0;
+
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) ->int {
```
Extra space after `WTF::switchOn(`

Also should probably use `-> size_t`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) ->int {
+        return CCECGetKeySize(options.get());
+        },
+        [](const Vector<uint8_t>& options) -> int {
```
Should probably use `-> size_t`.

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+    auto publicKey = CryptoKeyEC::create(identifier, curve, CryptoKeyType::Public, Vector<uint8_t>(ccPublicKey), true, usages);
+    auto privateKey = CryptoKeyEC::create(identifier, curve, CryptoKeyType::Private, Vector<uint8_t>(ccPrivateKey), extractable, usages);
     return CryptoKeyPair { WTFMove(publicKey), WTFMove(privateKey) };
+    
```
Extra line to remove.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+        if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data(), keyData.size(), ccECKeyPublic, &ccPublicKey))
+            return nullptr;
+        return create(identifier, curve, CryptoKeyType::Public, std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>(ccPublicKey), extractable, usages);
+        
```
extra line to remove.

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
-    return create(identifier, curve, CryptoKeyType::Public, PlatformECKeyContainer(ccPublicKey), extractable, usages);
+    if (usages & CryptoKeyUsageSign)
+        return create(identifier, curve, CryptoKeyType::Private, Vector<uint8_t>(keyData), extractable, usages);
+    return create(identifier, curve, CryptoKeyType::Public, Vector<uint8_t>(keyData), extractable, usages);
```
Seems this could be written like this:
```
auto keyType = usages & CryptoKeyUsageSign ? CryptoKeyType::Private : CryptoKeyType::Public;
return create(identifier, curve, keyType, Vector<uint8_t>(keyData), extractable, usages);
```
to avoid duplication.

angelaizg replied:
changed

youennf replied:
Chris comment does not seem addressed.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+    if (usages & CryptoKeyUsageSign)
+        return create(identifier, curve, CryptoKeyType::Private, Vector<uint8_t>(keyData), extractable, usages);
+    return create(identifier, curve, CryptoKeyType::Public, Vector<uint8_t>(keyData), extractable, usages);
+        
```
Extra line to remove.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
-        return { };
-    return result;
+    // Per Section 2.3.4 of http://www.secg.org/sec1-v2.pdf
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) -> Vector<uint8_t> {
```
Extra space after `switchOn(`

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+        if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, result.data(), &size, ccECKeyPublic, options.get()) || size != expectedSize))
+            return { };
+        return result;
+        },
```
Indentation issue

angelaizg replied:
changed

youennf replied:
This indentation issue seems still there

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +96,94 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+            return { };
+        return result;
+        },
+        [&](const Vector<uint8_t>& options) -> Vector<uint8_t> {
```
The content of `options` isn't used at all here, You are just returning a result vector with `expectedSize` items that are all 0.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -143,8 +195,8 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPublic(CryptoAlgorithmIdentifi
         return nullptr;
-
-    return create(identifier, curve, CryptoKeyType::Public, PlatformECKeyContainer(ccPublicKey), extractable, usages);
+    
```
Bad spacing change. Please change your IDE settings if it is your IDE doing this.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -164,37 +216,40 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPrivate(CryptoAlgorithmIdentif
-    switch (type()) {
-    case CryptoKeyType::Public:
-        if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, result.data(), &size, ccECKeyPublic, m_platformKey.get())))
+    return WTF::switchOn( m_platformKey, [&result, &size, this, &publicKeySize, &privateKeySize, &jwk, &keySizeInBytes](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options)mutable -> bool {
```
Extra space after `switchOn(`

cdumez replied:
Why not just capture `[&]` ?

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -164,37 +216,40 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPrivate(CryptoAlgorithmIdentif
+            jwk.d = base64URLEncodeToString(result.data() + publicKeySize, keySizeInBytes);
+        return true;
+    }, [&](const Vector<uint8_t>&options) -> bool {
+        return options.size()>0;
```
Missing spaces around `>`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +313,51 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-    if (doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index)) {
-        if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
+    if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index)) {
+        if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index+1, keyData.size() - index-1, ccECKeyPublic, &ccPublicKey))
```
Missing spaces around `+` and `-`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +313,51 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-
-    return result;
+    auto keyBytesdata = keyBytes.data();
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable->Vector<uint8_t> {
```
Extra space after `switchOn(`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +313,51 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-    return result;
+    auto keyBytesdata = keyBytes.data();
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable->Vector<uint8_t> {
+        if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, keyBytesdata , &keySize, ccECKeyPublic, options.get()) || keySize != expectedKeySize))
```
keyBytesdata -> keyBytesData?

Also, you have an extra space after keyBytesdata.

angelaizg replied:
changed to auto&

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +313,51 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+        result.append(keyBytes.data(), keyBytes.size());
+        return result;
+    }, [&](const Vector<uint8_t>&options) -> Vector<uint8_t> {
+        
```
extra blank line

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +313,51 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+    }, [&](const Vector<uint8_t>&options) -> Vector<uint8_t> {
+        
+        return options;
+
```
extra blank line

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +313,51 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+        result.append(InitialOctet);
+        result.append(keyBytes.data(), keyBytes.size());
+        return result;
+    }, [&](const Vector<uint8_t>&options) -> Vector<uint8_t> {
```
missing space before `options`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +429,60 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
-    if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, keyBytes.data(), &keySize, ccECKeyPrivate, m_platformKey.get()) || keySize != expectedKeySize))
-        return { };
+    Vector<uint8_t> result;
+    return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable ->Vector<uint8_t> {
```
extra space after `switchOn(`...

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +429,60 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
+        // VERSION + SEQUENCE + length(1) + OID id-ecPublicKey + OID secp256r1/OID secp384r1/OID secp521r1 + OCTET STRING + length(?) + privateKeySize
+        size_t totalSize = sizeof(Version) + sizeof(IdEcPublicKey) + oidSize + bytesNeededForEncodedLength(privateKeySize) + privateKeySize + 3;
+
+
```
Extra blank line.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +429,60 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
+        return result;
+    }, [&](const Vector<uint8_t>&options) -> Vector<uint8_t> {
+        return options;
+
```
extra blank line.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +429,60 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
+        result.append(keyBytes.data(), publicKeySize - 1);
+
+        return result;
+    }, [&](const Vector<uint8_t>&options) -> Vector<uint8_t> {
```
missing space before `options`.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2017-2019 Apple Inc. All rights reserved.
```
Is this moved code? why the old date?

angelaizg replied:
changed to 2022

cdumez commented on Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h:
```diff
@@ -0,0 +1,59 @@
+
+namespace WebCore {
+
+class CryptoAlgorithmEd25519Params final : public CryptoAlgorithmParameters {
```
Please use a struct instead of a class. 

Having a class with a public data member doesn't look good.

cdumez replied:
Also, you can drop the public in `public CryptoAlgorithmParameters`

angelaizg replied:
I was following the same structure, all of those type of files use a class, so I went with that schema.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h:
```diff
@@ -0,0 +1,59 @@
+namespace WebCore {
+
+class CryptoAlgorithmEd25519Params final : public CryptoAlgorithmParameters {
+public:
```
Not needed once you switch to using a struct.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h:
```diff
@@ -0,0 +1,59 @@
+SPECIALIZE_TYPE_TRAITS_CRYPTO_ALGORITHM_PARAMETERS(Ed25519Params)
+
+#endif // ENABLE(WEB_CRYPTO)
+
```
Too many blank lines.

angelaizg replied:
changed

cdumez commented on Source/WebCore/crypto/parameters/Ed25519Params.idl:
```diff
@@ -0,0 +1,31 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+ typedef (object or DOMString) HashAlgorithmIdentifier;
```
Missing blank line before this.

cdumez commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -5684,7 +5684,7 @@
 		63C32C271E98119000699BD0 /* _WKGeolocationPositionInternal.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = _WKGeolocationPositionInternal.h; sourceTree = "<group>"; };
 		63FABE191E970D65003011D5 /* _WKGeolocationCoreLocationProvider.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = _WKGeolocationCoreLocationProvider.h; sourceTree = "<group>"; };
-		650994DF28C942B7006EE6DB /* com.apple.WebKit.WebContent.Crashy.xpc */ = {isa = PBXFileReference; explicitFileType = "wrapper.xpc-service"; includeInIndex = 0; path = com.apple.WebKit.WebContent.Crashy.xpc; sourceTree = BUILT_PRODUCTS_DIR; };
+		650994DF28C942B7006EE6DB /* .xpc */ = {isa = PBXFileReference; explicitFileType = "wrapper.xpc-service"; includeInIndex = 0; path = .xpc; sourceTree = BUILT_PRODUCTS_DIR; };
```
Bad changes in this file. Please revert all changes to this file.

panva reviewed (commented):
https://wicg.github.io/webcrypto-secure-curves/ is the implementation already available in Deno (partial), Node.js (complete). It is also actively being worked on in chromium (https://bugs.chromium.org/p/chromium/issues/detail?id=1370697).

This webkit implementation should also follow this WICG draft and its already good WPT coverage.

panva commented on LayoutTests/crypto/subtle/ed25519-import-raw-key.html:
```diff
@@ -0,0 +1,39 @@
+    shouldBe("publicKey.type", "'public'");
+    shouldBe("publicKey.extractable", "true");
+    shouldBe("publicKey.algorithm.name", "'Ed25519'");
+    shouldBe("publicKey.algorithm.namedCurve", "'Curve25519'");
```
Neither key generation parameters nor the key algorithm should need/have a namedCurve property. This applies to all added CFRG curve algorithms.

angelaizg replied:
What are the CFRG algorithms ?

panva replied:
What I mean is that my comment applies to both Ed25519 and X25519 that you're actively working on.

youennf commented on LayoutTests/crypto/subtle/ed25519-import-raw-key.html:
```diff
@@ -0,0 +1,37 @@
+<script src="../../resources/js-test-post.js"></script>
+</body>
+</html>
+
```
unnecessary line

youennf commented on LayoutTests/crypto/subtle/ed25519-serialize.html:
```diff
@@ -0,0 +1,38 @@
+</body>
+</html>
+
+
```
unnecessary lines

angelaizg replied:
done

youennf commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,105 @@
+
+typedef uint8_t ccec25519signature[64];
+
+// https://codesearch.apple.com/result/corecrypto-1218.120.10/ccsha2%2Fcorecrypto%2Fccsha2.h?query=corecrypto%2Fccsha2.h&resultType=fileResults
```
To remove probably

angelaizg replied:
done

youennf commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,105 @@
+/* SHA512 */
+#define CCSHA512_BLOCK_SIZE  128
+#define    CCSHA512_OUTPUT_SIZE  64
+#define    CCSHA512_STATE_SIZE   64
```
unnecessary spaces.

angelaizg replied:
done

youennf commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,105 @@
+// https://codesearch.apple.com/result/corecrypto-1218.120.10/ccsha2%2Fcorecrypto%2Fccsha2.h?query=corecrypto%2Fccsha2.h&resultType=fileResults
+const struct ccdigest_info *ccsha512_di(void);
+#define ccoid_sha512 ((unsigned char *)"\x06\x09\x60\x86\x48\x01\x65\x03\x04\x02\x03")
+#define ccoid_sha512_len 11
```
This seems unused in this patch, if so, let's remove everything that is not used.

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -233,6 +241,11 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
             break;
         }
+        case CryptoAlgorithmIdentifier::Ed25519: {
+            auto result = makeUnique<CryptoAlgorithmParameters>();
```
Shouldn't we use params somehow?
If we should do makeUnique<CryptoAlgorithmParameters>(params), why are the test passing?

angelaizg replied:
Because there are no parameters defined for them

youennf replied:
I do not think this is correct to do so, we are probably losing CryptoAlgorithmParameters::name and CryptoAlgorithmParameters::identifier. This seems error prone especially since other code path are doing `makeUnique<CryptoAlgorithmParameters>(params);` like for CryptoAlgorithmIdentifier::RSAES_PKCS1_v1_5 case.

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -317,6 +330,11 @@ static ExceptionOr<std::unique_ptr<CryptoAlgorithmParameters>> normalizeCryptoAl
         }
+        case CryptoAlgorithmIdentifier::Ed25519: {
+            auto result = makeUnique<CryptoAlgorithmParameters>();
+            result->identifier = identifier.value();
```
Ditto.

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -448,6 +466,7 @@ static void normalizeJsonWebKey(JsonWebKey& webKey)
 // should be resolved by adopting a standards compliant std::variant (see https://webkit.org/b/175583)
 static std::optional<KeyData> toKeyData(SubtleCrypto::KeyFormat format, SubtleCrypto::KeyDataVariant&& keyDataVariant, Ref<DeferredPromise>& promise)
 {
+    WTFLogAlways("toKeyDataSubtleCrypto");
```
To remove

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -690,6 +710,7 @@ void SubtleCrypto::verify(JSC::JSGlobalObject& state, AlgorithmIdentifier&& algo
     auto callback = [index, weakThis](bool result) mutable {
         if (auto promise = getPromise(index, weakThis))
             promise->resolve<IDLBoolean>(result);
+
```
To remove

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,209 @@
+
+void CryptoAlgorithmEd25519::generateKey(const CryptoAlgorithmParameters&, bool extractable, CryptoKeyUsageBitmap usages, KeyOrKeyPairCallback&& callback, ExceptionCallback&& exceptionCallback, ScriptExecutionContext&)
+{
+    
```
Line unneeded

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,209 @@
+    
+    auto result = CryptoKeyEC::generatePair(CryptoAlgorithmIdentifier::Ed25519, "Curve25519"_s, extractable, usages);
+    if (result.hasException()) {
+        exceptionCallback(result.releaseException().code());
```
This does not need to be fixed right now, but it is a bit sad that we are dropping the message here.
Probably ExceptionCallback should take an Exception instead of an ExceptionCode.

angelaizg replied:
will file a bug for this task to be done on another patch

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,209 @@
+    
+    dispatchOperationInWorkQueue(workQueue, context, WTFMove(callback), WTFMove(exceptionCallback),
+        [parameters = crossThreadCopy(downcast<CryptoAlgorithmEd25519Params>(parameters)), key = WTFMove(key), data = WTFMove(data)] {
+            return platformSign( downcast<CryptoKeyEC>(key.get()), data);
```
space unneeded

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,209 @@
+            return platformVerify(downcast<CryptoKeyEC>(key.get()), signature, data);
+        });
+}
+void CryptoAlgorithmEd25519::importKey(CryptoKeyFormat format, KeyData&& data, const CryptoAlgorithmParameters&, bool extractable, CryptoKeyUsageBitmap usages, KeyCallback&& callback, ExceptionCallback&& exceptionCallback)
```
This method seems very similar to CryptoAlgorithmECDSA::importKey.
Is there a way to share code between the two.
Ditto for the other methods.

angelaizg replied:
the structure of those functions are the same for all the algorithms, a switch in which the cases of jwk, spki or raw are treated. In each different algorithm the way it is  treated is different there are some cases that might be more similar than others but there are other cases very different, for example jwk. In any case maybe we could try to abstract them for all the algorithms but I don't think it makes much senses in this patch.

youennf commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h:
```diff
@@ -0,0 +1,56 @@
+    CryptoAlgorithmIdentifier identifier() const final;
+    
+    void generateKey(const CryptoAlgorithmParameters& , bool extractable, CryptoKeyUsageBitmap usages, KeyOrKeyPairCallback&& , ExceptionCallback&& , ScriptExecutionContext&);
+    void sign(const CryptoAlgorithmParameters&, Ref<CryptoKey>&&, Vector<uint8_t>&&, VectorCallback&&, ExceptionCallback&&, ScriptExecutionContext&, WorkQueue&) final;
```
We usually prepend the overriden methods with a comment with name of the base interface.
Maybe we could add `// CryptoAlgorithm` here.

angelaizg replied:
where do you mean? In the other algorithms couldn't see such thing either

youennf replied:
See for instance AudioContext.h as an example:
```
    // ActiveDOMObject
    const char* activeDOMObjectName() const final;
```

angelaizg replied:
That would mean adding comments to almost all the functions but to have a clearer code as in the other algorithms maybe it is better to leave it this way

youennf commented on Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp:
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ * Copyright (C) 2022 Metrological Group B.V.
+ * Copyright (C) 2022 Igalia S.L.
```
Is it normal?

angelaizg replied:
What do you mean?

youennf replied:
It is a new file so only Apple Copyright should be there.

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -89,7 +92,18 @@ class CryptoKeyEC final : public CryptoKey {
+    return options;
+    });
+    }
+#else
```
Code is not well formatted. Let's move it to Cpp file also.

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -89,7 +92,18 @@ class CryptoKeyEC final : public CryptoKey {
     NamedCurve namedCurve() const { return m_curve; }
     String namedCurveString() const;
+    
+#if OS(DARWIN) && !PLATFORM(GTK)
```
Why do we need OS(DARWIN)?
Should we just have PLATFORM(COCOA)?

angelaizg replied:
I it is to handle that case in linux platforms

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,75 @@
+
+ExceptionOr<Vector<uint8_t>> CryptoAlgorithmEd25519::platformSign(const CryptoKeyEC& key, const Vector<uint8_t>& data)
+{
+    return signEd25519( key.platformKey(), key.keySizeInBytes(), data);
```
space unneeded

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,75 @@
+
+ExceptionOr<bool> CryptoAlgorithmEd25519::platformVerify(const CryptoKeyEC& key, const Vector<uint8_t>& signature, const Vector<uint8_t>& data)
+{
+    return verifyEd25519( key.platformKey(), key.keySizeInBytes(), signature, data);
```
space unneeded

angelaizg replied:
where ?

panva replied:
```
return verifyEd25519( key.platformKey()
                     ^ there
```

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,75 @@
+
+static ExceptionOr<Vector<uint8_t>> signEd25519(const PlatformECKey key, size_t, const Vector<uint8_t>& data)
+{   
+    auto sk = std::get<Vector<uint8_t>>(key);
```
This is probably a memory copy, we probably want to take a ref here.

angelaizg replied:
do you mean setting a pointer instead?

angelaizg replied:
do you suggest using a ptr ?

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,75 @@
+    ccec25519signature newSignature;
+    cced25519_sign(di, newSignature, data.size(), data.data(), pk, sk.data());
+    Vector<uint8_t> sg(newSignature, 64);
+    return WTFMove(sg);
```
Next two lines can be made as a one liner, we usually do not need return WTFMove().

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -23,13 +23,16 @@
  */
 
+#include "config.h"
 #include "config.h"
```
Not needed

angelaizg replied:
done

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -23,13 +23,16 @@
 #include "CommonCryptoDERUtilities.h"
 #include "JsonWebKey.h"
+#include <pal/spi/cocoa/CoreCryptoSPI.h>
+#include <stdlib.h>
```
Do we need stdlib.h include?

angelaizg replied:
I included it since there are types as size_t defined there.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -90,48 +97,84 @@ static constexpr bool doesFieldElementMatchNamedCurve(CryptoKeyEC::NamedCurve cu
+        auto publicKey = CryptoKeyEC::create(identifier, curve, CryptoKeyType::Public, std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>(ccPublicKey), true, usages);
+        auto privateKey = CryptoKeyEC::create(identifier, curve, CryptoKeyType::Private, std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>(ccPrivateKey), extractable, usages);
+        return CryptoKeyPair { WTFMove(publicKey), WTFMove(privateKey) };
+        
```
line unneeded.

angelaizg replied:
done

angelaizg replied:
done

youennf reviewed (commented):
Some comments below.
Please also take care of preexisting comments from Chris.
The commit message needs also to be improved: please describe what you are doing in this patch with as much detail as possible like the design choices..., especially since this is a big patch. The list of files changed in the commit messages is also wrong.

Commit:
Support for Curve 25519 - Algorithm Ed25519
https://bugs.webkit.org/show_bug.cgi?id=246145
rdar://100401588

Reviewed by NOBODY (OOPS!).

Adding support for the Curve 25519 (doing the necessary modifications to be able to work with such type  of curve), specifically the implementation of the Ed25519 algorithm, as in https://www.rfc-editor.org/rfc/rfc7748. The current patch adds support  to generate keys, import, export (in raw format for now), sign and verify and tests that make use of such operations. Following the specs in https://wicg.github.io/webcrypto-secure-curves/#ed25519.

* LayoutTests/crypto/subtle/ed25519-generate-export-key-raw-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-generate-export-key-raw.html: Added.
* LayoutTests/crypto/subtle/ed25519-generate-key-sing-verify-Curve25519-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-generate-key-sing-verify-Curve25519.html: Added.
* LayoutTests/crypto/subtle/ed25519-import-raw-key-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-import-raw-key.html: Added.
* LayoutTests/crypto/subtle/ed25519-serialize-expected.txt: Added.
* LayoutTests/crypto/subtle/ed25519-serialize.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h: Added.
(cccurve25519_make_priv):
(cccurve25519_make_pub):
(cccurve25519_make_key_pair):
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::read):
* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
* Source/WebCore/crypto/CryptoAlgorithmParameters.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::normalizeCryptoAlgorithmParameters):
(WebCore::isSupportedExportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmECDH.cpp:
(WebCore::CryptoAlgorithmECDH::generateKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmECDSA.cpp:
(WebCore::CryptoAlgorithmECDSA::generateKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp: Copied from Source/WebCore/crypto/algorithms/CryptoAlgorithmECDSA.cpp.
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
(WebCore::CryptoAlgorithmEd25519::create):
(WebCore::CryptoAlgorithmEd25519::identifier const):
(WebCore::CryptoAlgorithmEd25519::generateKey):
(WebCore::CryptoAlgorithmEd25519::sign):
(WebCore::CryptoAlgorithmEd25519::verify):
(WebCore::CryptoAlgorithmEd25519::importKey):
(WebCore::CryptoAlgorithmEd25519::exportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h: Added.
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmEd25519GCrypt.cpp: Copied from Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp.
(WebCore::gcryptSign):
(WebCore::gcryptVerify):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmRegistryGCrypt.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/gcrypt/CryptoKeyECGCrypt.cpp:
(WebCore::curveName):
(WebCore::curveIdentifier):
(WebCore::curveSize):
(WebCore::curveUncompressedFieldElementSize):
* Source/WebCore/crypto/keys/CryptoKeyEC.cpp:
(WebCore::toNamedCurve):
(WebCore::CryptoKeyEC::exportJwk const):
(WebCore::CryptoKeyEC::namedCurveString const):
(WebCore::CryptoKeyEC::isValidECAlgorithm):
(WebCore::CryptoKeyEC::algorithm const):
* Source/WebCore/crypto/keys/CryptoKeyEC.h:
* Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp:
(WebCore::CryptoAlgorithmECDH::platformDeriveBits):
* Source/WebCore/crypto/mac/CryptoAlgorithmECDSAMac.cpp:
(WebCore::signECDSA):
(WebCore::verifyECDSA):
* Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp: Added.
(WebCore::signEd25519):
(WebCore::verifyEd25519):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/mac/CryptoAlgorithmHKDFMac.cpp:
* Source/WebCore/crypto/mac/CryptoAlgorithmRegistryMac.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
(WebCore::keySizeInBitsFromNamedCurve):
(WebCore::CryptoKeyEC::keySizeInBits const):
(WebCore::CryptoKeyEC::platformSupportedCurve):
(WebCore::CryptoKeyEC::platformGeneratePair):
(WebCore::CryptoKeyEC::platformImportRaw):
(WebCore::CryptoKeyEC::platformExportRaw const):
(WebCore::CryptoKeyEC::platformImportJWKPublic):
(WebCore::CryptoKeyEC::platformImportJWKPrivate):
(WebCore::CryptoKeyEC::platformAddFieldElements const):
(WebCore::getOID):
(WebCore::CryptoKeyEC::platformImportSpki):
(WebCore::CryptoKeyEC::platformExportSpki const):
(WebCore::CryptoKeyEC::platformImportPkcs8):
(WebCore::CryptoKeyEC::platformExportPkcs8 const):
* Source/WebCore/crypto/openssl/CryptoAlgorithmEd25519OpenSSL.cpp: Copied from Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp.
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/openssl/CryptoAlgorithmRegistryOpenSSL.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h: Copied from Source/WebCore/crypto/CryptoAlgorithmIdentifier.h.
* Source/WebCore/crypto/parameters/Ed25519Params.idl: Copied from Source/WebCore/crypto/CryptoAlgorithmIdentifier.h.
* Source/WebCore/css/StylePropertyShorthand.h:

youennf commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -0,0 +1,93 @@
+    cccurve25519(pk, sk, CC_WIDE_NULL);
+}
+
+inline void cccurve25519_make_key_pair(const struct ccdigest_info *, struct ccrng_state *rng, ccec25519pubkey pk, ccec25519secretkey sk)
```
I do not see this function being used anywhere, why is it there?
As I said in a past review, please validate all declarations in this file are actually used.

angelaizg replied:
This function is used in the second patch, for the X25519 so I added the whole file for that

youennf commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -3000,6 +3004,9 @@ class CloneDeserializer : CloneBase {
             result = CryptoAlgorithmIdentifier::ECDH;
             break;
+        case CryptoAlgorithmIdentifierTag::Ed25519:
+            result  = CryptoAlgorithmIdentifier::Ed25519;
```
As I said in a past review, it would be good to have a test for key serialisation.
That could be covered by using postMessage with a Ed25519 key.

angelaizg replied:
I added a tests searializationt that imports and exports the key, is that what you meant ? Maybe I did get you wrong

youennf commented on Source/WebCore/crypto/gcrypt/CryptoAlgorithmRegistryGCrypt.cpp:
```diff
@@ -57,6 +58,7 @@ void CryptoAlgorithmRegistry::platformRegisterAlgorithms()
     registerAlgorithm<CryptoAlgorithmAES_KW>();
     registerAlgorithm<CryptoAlgorithmECDH>();
     registerAlgorithm<CryptoAlgorithmECDSA>();
+    registerAlgorithm<CryptoAlgorithmEd25519>();
```
Given there is no platform support, it might be better to not register this algorithm

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -89,7 +92,18 @@ class CryptoKeyEC final : public CryptoKey {
     NamedCurve namedCurve() const { return m_curve; }
     String namedCurveString() const;
+    
+#if PLATFORM(COCOA) && !PLATFORM(GTK)
```
This can be replaced by `#if PLATFORM(COCOA)`

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -89,7 +92,18 @@ class CryptoKeyEC final : public CryptoKey {
+    PlatformECKey platformKey() const { return WTF::switchOn( m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) -> PlatformECKey {
+        return options.get();
+            
+        }, [&](const Vector<uint8_t>& options) -> PlatformECKey {
```
This is not correctly indented, there is an unnecessary line.
Also, we tend to not add multiline methods like this in headers.

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -33,14 +33,16 @@
 #include "CommonCryptoUtilities.h"
+#include <wtf/Vector.h>
+
+using PlatformECKey = std::variant<CCECCryptorRef, Vector<uint8_t>>;
```
Since CCECCryptorRef is just a ref, I wonder whether we should try to return :
`std::variant<CCECCryptorRef, std::reference_wrapper<Vector<uint8_t>>>`.
That will remove the memory allocation/copy in case of variant be a vector.

Another approach would be to remove platformKey() and introduce platformKeyAsCCECCryptoRef() and platformKeyAsByteArray().
These getters would return either a `CCECCryptorRef` or a `const Vector<uint8_t>&`

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/keys/CryptoKeyEC.h:
```diff
@@ -89,7 +92,18 @@ class CryptoKeyEC final : public CryptoKey {
+        return options.get();
+            
+        }, [&](const Vector<uint8_t>& options) -> PlatformECKey {
+        return options;
```
This creates an unnecessary memory allocation/copy probably

angelaizg replied:
what would be the alternative
/

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp:
```diff
@@ -38,7 +38,9 @@ std::optional<Vector<uint8_t>> CryptoAlgorithmECDH::platformDeriveBits(const Cry
     Vector<uint8_t> derivedKey(baseKey.keySizeInBytes()); // Per https://tools.ietf.org/html/rfc6090#section-4.
     size_t size = derivedKey.size();
-    if (!CCECCryptorComputeSharedSecret(baseKey.platformKey(), publicKey.platformKey(), derivedKey.data(), &size))
+    CCECCryptorRef cceccbasekey = std::get<CCECCryptorRef>(baseKey.platformKey());
```
If we were to add typed getters, we would write it as:
`auto cceccbasekey = baseKey.platformKeyAs CCECCryptorRef()`

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmECDHMac.cpp:
```diff
@@ -38,7 +38,9 @@ std::optional<Vector<uint8_t>> CryptoAlgorithmECDH::platformDeriveBits(const Cry
     size_t size = derivedKey.size();
-    if (!CCECCryptorComputeSharedSecret(baseKey.platformKey(), publicKey.platformKey(), derivedKey.data(), &size))
+    CCECCryptorRef cceccbasekey = std::get<CCECCryptorRef>(baseKey.platformKey());
+    CCECCryptorRef cceccpublickey = std::get<CCECCryptorRef>(publicKey.platformKey());
```
here and above, we can use `auto`

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,74 @@
+#include <stdio.h>
+#include <string.h>
+#include <wtf/Assertions.h>
+#include <wtf/text/Base64.h>
```
Are this include and others like stdio/string needed?

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,74 @@
+
+namespace WebCore {
+
+static ExceptionOr<Vector<uint8_t>> signEd25519(const PlatformECKey key, size_t, const Vector<uint8_t>& data)
```
We are not using size_t parameter, let's remove it.
Also, why are we returning ExceptionOr<Vector>, it seems we can return Vector directly

angelaizg replied:
all the algorithms are written that way, following the same pattern

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,74 @@
+{   
+    auto& sk = std::get<Vector<uint8_t>>(key);
+    ccec25519pubkey pk;
+    const struct ccdigest_info* di = ccsha512_di();
```
`auto*`

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,74 @@
+{
+    if (signature.size() != keyLengthInBytes * 2)
+        return false;
+    const struct ccdigest_info* di = ccsha512_di();
```
auto*

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,74 @@
+    return Vector<uint8_t>(newSignature, 64);
+}
+
+static ExceptionOr<bool> verifyEd25519(const PlatformECKey key, size_t keyLengthInBytes, const Vector<uint8_t>& signature, const Vector<uint8_t> data)
```
It seems this function could return bool instead of ExceptionOr<bool>.

youennf commented on Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Mac.cpp:
```diff
@@ -0,0 +1,74 @@
+static ExceptionOr<bool> verifyEd25519(const PlatformECKey key, size_t keyLengthInBytes, const Vector<uint8_t>& signature, const Vector<uint8_t> data)
+{
+    if (signature.size() != keyLengthInBytes * 2)
+        return false;
```
Should we return false or an exception?

angelaizg replied:
False, it just gives information that it does not verify it because the length of the key is not what it should be

youennf commented on Source/WebCore/crypto/openssl/CryptoAlgorithmEd25519OpenSSL.cpp:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2022 Sony Interactive Entertainment Inc.
```
This does not seem correct

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -144,7 +182,7 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPublic(CryptoAlgorithmIdentifi
         return nullptr;
 
-    return create(identifier, curve, CryptoKeyType::Public, PlatformECKeyContainer(ccPublicKey), extractable, usages);
+    return create(identifier, curve, CryptoKeyType::Public, std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>(ccPublicKey), extractable, usages);
```
std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter> is used in multiple place and is pretty long.
Let's name this type.

angelaizg replied:
named it as PlatformECKeyContainer_first

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -164,37 +202,41 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPrivate(CryptoAlgorithmIdentif
     size_t publicKeySize = keySizeInBytes * 2 + 1; // 04 + X + Y per Section 2.3.4 of http://www.secg.org/sec1-v2.pdf
     size_t privateKeySize = keySizeInBytes * 3 + 1; // 04 + X + Y + D
-
+    
```
Change is unneeded

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -164,37 +202,41 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPrivate(CryptoAlgorithmIdentif
+        if (size > publicKeySize)
+            jwk.d = base64URLEncodeToString(result.data() + publicKeySize, keySizeInBytes);
+        return true;
+        }, [&](const Vector<uint8_t>&options) -> bool {
```
missing space

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -164,37 +202,41 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportJWKPrivate(CryptoAlgorithmIdentif
+            jwk.d = base64URLEncodeToString(result.data() + publicKeySize, keySizeInBytes);
+        return true;
+        }, [&](const Vector<uint8_t>&options) -> bool {
+            return options.size() > 0;
```
Let's write it as:
`return !options.isEmpty();`

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +300,49 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-        if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
+        if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
             return nullptr;
-        
```
Unneeded change

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +300,49 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-    if (doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index)) {
-        if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
+    if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index)) {
+        if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index + 1, keyData.size() - index - 1, ccECKeyPublic, &ccPublicKey))
```
These changes are probably unrelated.

angelaizg replied:
It is related since adding a new case there, needed such checks otherwise was throwing an error.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +300,49 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
     } else {
-        ++index;
-        if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
+        if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
```
Ditto here

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +300,49 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-
-    return result;
+    auto keyBytesData = keyBytes.data();
+    return WTF::switchOn(m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable->Vector<uint8_t> {
```
missing space

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +300,49 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+    return WTF::switchOn(m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable->Vector<uint8_t> {
+        if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, keyBytesData , &keySize, ccECKeyPublic, options.get()) || keySize != expectedKeySize))
+            return { };
+        // The following addes SPKI header to a raw EC public key.
```
s/addes/adds/

angelaizg replied:
I don't understand this comment

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -254,49 +300,49 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+
+        // SEQUENCE + length(1) + OID id-ecPublicKey + OID secp256r1/OID secp384r1/OID secp521r1 + BIT STRING + length(?) + InitialOctet + Key size
+        size_t totalSize = sizeof(IdEcPublicKey) + oidSize + bytesNeededForEncodedLength(keySize + 1) + keySize + 4;
+        result.reserveInitialCapacity(totalSize + bytesNeededForEncodedLength(totalSize) + 1);
```
reserveInitialCapacity is usually interesting with uncheckedAppend if possible.
But we need to be sure of the exact init size.
Since this is preexisting code this is probably fine to keep as is but we could do this as a follow-up

angelaizg replied:
Ok, we can keep it then

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -355,12 +401,11 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportPkcs8(CryptoAlgorithmIdentifier i
     if (!doesUncompressedPointMatchNamedCurve(curve, keyBinary.size()))
         return nullptr;
     keyBinary.append(keyData.data() + privateKeyPos, privateKeySize);
-
```
unneeded change

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +414,58 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
-        return { };
+    Vector<uint8_t> result;
+    return WTF::switchOn(m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable ->Vector<uint8_t> {
+
```
line unneeded

angelaizg replied:
changed

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +414,58 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
-    if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, keyBytes.data(), &keySize, ccECKeyPrivate, m_platformKey.get()) || keySize != expectedKeySize))
-        return { };
+    Vector<uint8_t> result;
+    return WTF::switchOn(m_platformKey, [&](const std::unique_ptr<typename std::remove_pointer<CCECCryptorRef>::type, WebCore::CCECCryptorRefDeleter>& options) mutable ->Vector<uint8_t> {
```
missing space

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +414,58 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
     size_t keySize = keyBytes.size();
-    if (UNLIKELY(CCECCryptorExportKey(kCCImportKeyBinary, keyBytes.data(), &keySize, ccECKeyPrivate, m_platformKey.get()) || keySize != expectedKeySize))
-        return { };
+    Vector<uint8_t> result;
```
result does not need to be defined here, since we return switchOn.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -369,51 +414,58 @@ Vector<uint8_t> CryptoKeyEC::platformExportPkcs8() const
+        size_t totalSize = sizeof(Version) + sizeof(IdEcPublicKey) + oidSize + bytesNeededForEncodedLength(privateKeySize) + privateKeySize + 3;
+        
+        result.reserveInitialCapacity(totalSize + bytesNeededForEncodedLength(totalSize) + 1);
+        result.append(SequenceMark);
```
Ditto with reserveInitialCapacity and uncheckedAppend(span<uint8_t>).

Since this is preexisting code this is probably fine to keep as is but we could do this as a follow-up

angelaizg replied:
ok

youennf commented on Source/WebCore/crypto/parameters/CryptoAlgorithmEd25519Params.h:
```diff
@@ -0,0 +1,53 @@
+
+    CryptoAlgorithmEd25519Params isolatedCopy() const
+    {
+        CryptoAlgorithmEd25519Params result;
```
Is this correct?
Should we copy all CryptoAlgorithmParameters members including `name`?
